config/env_config.go

```diff
@@ -0,0 +1,431 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package config defines the API Runtime Control config and provides
+// the config loading and validation functions.
+
+package config
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/apigee/apigee-remote-service-golib/v2/errorset"
+	"gopkg.in/yaml.v3"
+)
+
+// validateEnvConfigs checks if there are
+//   * environment configs with the same ID,
+//   * API configs under the same environment config with the same ID,
+//   * JWT authentication requirement under the same API or operation with the same name
+// and report them as errors
+func (c *Config) validateEnvConfigs() error {
```

Comment:
JFYI we'll need very similar validation logic in the control plane, so let's think about how we can scale this sort of development (don't want to write the code twice).

Reply:
If the validation operates on those two schemas, I am not sure how we can avoid repeating at least some of it. Alternatively and not sure if desired, we could generate this and then validate it in the control plane. I would just need to modify the signature of this method (if needed) and export it.

Reply:
Yea this is really just a reminder to think about how we can structure code for minimal redundancy, so focus more on structural correctness (no duplicate IDs) versus semantic correctness e.g. that string values use valid syntax, have reasonable lengths, etc.

And yes I agree we will want to run this validator on the generated config in the control plane, the same as we will do for Envoy as a last-ditch effort to avoid a bad rollout. However the error messages will not be even close to as useful as errors generated in the original input validation layer since they won't have any user-facing context.

Reply:
Yes the current validation simply checks for all the duplication. We can add more checks as needed as it evolves.

In the latest commit, I have exported this function.

Reply:
Can we create a new issue for this if needed and move this forward?

Reply:
Created #205.

Reply:
(Or to put another way: @josheinhorn, are you -1 on merging this?)